app/src/main/java/protect/card_locker/ManageGroupActivity.java

```diff
@@ -0,0 +1,264 @@
+                    toast.show();
+                    return;
+                }
+            }
+
+            mAdapter.commitToDatabase();
+            Toast toast = Toast.makeText(getApplicationContext(), R.string.group_updated, Toast.LENGTH_SHORT);
+            if(!currentGroupName.trim().equals(mGroup._id)){
+                mDB.updateGroup(mGroup._id, currentGroupName.trim());
+            }
+            toast.show();
+            finish();
+        });
+        mHelpText.setText(getResources().getText(R.string.noGiftCardsGroup));
+    }
+
+    private void checkIfGroupNameIsInUse(){
+        mGroupNameNotInUse = false;
+        String currentGroupName = mGroupNameText.getText().toString();
+        if (!mGroup._id.equals(currentGroupName.trim())) {
+            Group group = mDB.getGroup(currentGroupName.trim());
+            if (group != null) {
+                mGroupNameNotInUse = false;
+                mGroupNameText.setError(getResources().getText(R.string.group_name_already_in_use));
+            } else {
+                mGroupNameNotInUse = true;
+            }
+        }
+    }
+
+    @Override
+    protected void onResume()
+    {
+        super.onResume();
+
+        setTitle(getString(R.string.editGroup, mGroup._id));
+
+        if (mCurrentGroupName == null){
+            mGroupNameText.setText(mGroup._id);
+        }else{
+            mGroupNameText.setText(mCurrentGroupName);
+            checkIfGroupNameIsInUse();
+        }
+
+        mGroupNameText.addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+                mGroupNameText.setError(null);
+                checkIfGroupNameIsInUse();
+            }
+        });
+
+        if (mAdapterState != null){
+            mAdapter.importInGroupState(mAdapterState);
+        }
+        updateLoyaltyCardList();
+    }
+
+    private ArrayList<Integer> adapterStateToIntegerArray(){
+        ArrayList<Integer> ret = new ArrayList<>(mAdapterState.size() * 2);
+        for (Map.Entry<Integer, Boolean> entry : mAdapterState.entrySet()) {
+            ret.add(entry.getKey());
+            ret.add(entry.getValue()?1:0);
+        }
+        return ret;
+    }
+
+    private void integerArrayToAdapterState(ArrayList<Integer> in) {
+        boolean isKey = true;
+        int cardId = 0;
+        mAdapterState = new HashMap<>();
+        for (int value : in) {
+            if (isKey) {
+                cardId = value;
+            } else {
+                mAdapterState.put(cardId, value == 1);
+            }
+            isKey = !isKey;
+        }
```

Comment:
Instead of constantly flipping isKey, why not just loop per two?

```
for (int i = 0; i < in.size(); i += 2) {
    mAdapterState.put(in[i], in[i + 1] == 1);
}
```

Reply:
oh true it's an array list